dyno-recipes/src/main/java/com/netflix/dyno/recipes/lock/DynoLockClient.java

```diff
@@ -0,0 +1,369 @@
+package com.netflix.dyno.recipes.lock;
+
+import com.netflix.discovery.EurekaClient;
+import com.netflix.dyno.connectionpool.ConnectionPool;
+import com.netflix.dyno.connectionpool.ConnectionPoolMonitor;
+import com.netflix.dyno.connectionpool.HostSupplier;
+import com.netflix.dyno.connectionpool.TokenMapSupplier;
+import com.netflix.dyno.connectionpool.impl.ConnectionPoolConfigurationImpl;
+import com.netflix.dyno.contrib.ArchaiusConnectionPoolConfiguration;
+import com.netflix.dyno.contrib.DynoCPMonitor;
+import com.netflix.dyno.contrib.DynoOPMonitor;
+import com.netflix.dyno.jedis.DynoJedisClient;
+import com.netflix.dyno.jedis.DynoJedisUtils;
+import com.netflix.dyno.recipes.lock.command.CheckAndRunHost;
+import com.netflix.dyno.recipes.lock.command.ExtendHost;
+import com.netflix.dyno.recipes.lock.command.LockHost;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import redis.clients.jedis.Jedis;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+/**
+ * Client for acquiring locks similar to the redlock implementation https://redis.io/topics/distlock
+ */
```

Comment:
I think documenting some of the important implementation details here would be very helpful for users and debugging for developers.

The kind of details that would be useful are:
- The working of the common locking case
- What happens when one or more hosts are unreachable during acquire/extendLock()
- What happens when one or more hosts are unreachable during releaseLock()
- How we choose to reschedule extend() tasks (and why we choose to do it that way).
- Clock drift considerations.
- Possible edge case races we should be worried about.

Reply:
I've added some comments on the major points. Will add a few more in the wiki for this. Working on the wiki right now.